### ISSUE TITLE
Poetry environment broken after swapping package from PyPI release to git repo

### ISSUE THREAD

**Issue by LK** [python-poetry/poetry#7327]
- **Poetry version**: 1.3.1
- **Python version**: 3.7.3
- **OS version and name**: macOS 13.1
- **pyproject.toml**: https://gist.github.com/LK/3fcddd8f6cfe3ec00cfcc514e7700446

- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below.

## Issue

Repro steps:

1. `poetry lock`
2. `poetry install`
3. Swap the commented lines in the `pyproject.toml`, to point to the git repo instead of the PyPI release
4. `poetry lock`
5. `poetry install` or `poetry lock` fails with `No git repository was found at /Users/lennykhazan/Library/Caches/pypoetry/virtualenvs/api-server-apps-UDS0uKpv-py3.9/src/opencensus`.

The only way I've found to recover from this is to `poetry env remove --all` and then `poetry install` again, which seems to work.

<details>
<summary>Logs</summary>

```
❯ poetry install -vvv
Loading configuration file /Users/lennykhazan/Library/Preferences/pypoetry/config.toml
Using virtualenv: /Users/lennykhazan/Library/Caches/pypoetry/virtualenvs/api-server-apps-UDS0uKpv-py3.9

  Stack trace:

  18  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/application.py:327 in run
       325│
       326│             try:
     → 327│                 exit_code = self._run(io)
       328│             except BrokenPipeError:
       329│                 # If we are piped to another process, it may close early and send a

  17  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/console/application.py:190 in _run
       188│         self._load_plugins(io)
       189│
     → 190│         exit_code: int = super()._run(io)
       191│         return exit_code
       192│

  16  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/application.py:431 in _run
       429│             io.input.interactive(interactive)
       430│
     → 431│         exit_code = self._run_command(command, io)
       432│         self._running_command = None
       433│

  15  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/application.py:473 in _run_command
       471│
       472│         if error is not None:
     → 473│             raise error
       474│
       475│         return terminate_event.exit_code

  14  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/application.py:454 in _run_command
       452│
       453│         try:
     → 454│             self._event_dispatcher.dispatch(command_event, COMMAND)
       455│
       456│             if command_event.command_should_run():

  13  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/events/event_dispatcher.py:26 in dispatch
        24│
        25│         if listeners:
     →  26│             self._do_dispatch(listeners, event_name, event)
        27│
        28│         return event

  12  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/cleo/events/event_dispatcher.py:89 in _do_dispatch
        87│                 break
        88│
     →  89│             listener(event, event_name, self)
        90│
        91│     def _sort_listeners(self, event_name: str) -> None:

  11  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/console/application.py:322 in configure_installer_for_event
       320│             return
       321│
     → 322│         cls.configure_installer_for_command(command, event.io)
       323│
       324│     @staticmethod

  10  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/console/application.py:336 in configure_installer_for_command
       334│             poetry.pool,
       335│             poetry.config,
     → 336│             disable_cache=poetry.disable_cache,
       337│         )
       338│         installer.use_executor(poetry.config.get("experimental.new-installer", False))

   9  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/installation/installer.py:78 in __init__
        76│         self._installer = self._get_installer()
        77│         if installed is None:
     →  78│             installed = self._get_installed()
        79│
        80│         self._installed_repository = installed

   8  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/installation/installer.py:569 in _get_installed
       567│
       568│     def _get_installed(self) -> InstalledRepository:
     → 569│         return InstalledRepository.load(self._env)
       570│

   7  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/repositories/installed_repository.py:291 in load
       289│                     continue
       290│
     → 291│                 package = cls.create_package_from_distribution(distribution, env)
       292│
       293│                 if with_dependencies:

   6  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/repositories/installed_repository.py:141 in create_package_from_distribution
       139│                                 source_url,
       140│                                 source_reference,
     → 141│                             ) = cls.get_package_vcs_properties_from_path(src)
       142│                             break
       143│

   5  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/repositories/installed_repository.py:89 in get_package_vcs_properties_from_path
        87│         from poetry.vcs.git import Git
        88│
     →  89│         info = Git.info(repo=src)
        90│         return "git", info.origin, info.revision
        91│

   4  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/vcs/git/backend.py:174 in info
       172│     @classmethod
       173│     def info(cls, repo: Repo | Path | str) -> GitRepoLocalInfo:
     → 174│         return GitRepoLocalInfo(repo=repo)
       175│
       176│     @staticmethod

   3  <string>:2 in __init__
         1│

   2  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/vcs/git/backend.py:144 in __post_init__
       142│
       143│     def __post_init__(self, repo: Repo | Path | str) -> None:
     → 144│         repo = Git.as_repo(repo=repo) if not isinstance(repo, Repo) else repo
       145│         self.origin = Git.get_remote_url(repo=repo, remote="origin")
       146│         self.revision = Git.get_revision(repo=repo)

   1  ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/poetry/vcs/git/backend.py:152 in as_repo
       150│     @staticmethod
       151│     def as_repo(repo: Path | str) -> Repo:
     → 152│         return Repo(str(repo))
       153│
       154│     @staticmethod

  NotGitRepository

  No git repository was found at /Users/lennykhazan/Library/Caches/pypoetry/virtualenvs/api-server-apps-UDS0uKpv-py3.9/src/opencensus

  at ~/Library/Application Support/pypoetry/venv/lib/python3.7/site-packages/dulwich/repo.py:1127 in __init__
      1123│                     and os.path.isdir(os.path.join(root, REFSDIR))):
      1124│                 bare = True
      1125│             else:
      1126│                 raise NotGitRepository(
    → 1127│                     "No git repository was found at %(path)s" % dict(path=root)
      1128│                 )
      1129│
      1130│         self.bare = bare
      1131│         if bare is False:
```

</details>

**Comment by dimbleby** [comment]
duplicate #7206, which apparently was a duplicate of #6322

**Comment by dimbleby** [comment]
Not sure any of these actually are the same, though.

Looks to me as though the use of a `subdirectory` in a git dependency causes poetry to put code in the wrong place in the virtual environment: it all ends up in `.virtualenv/<whatever>/src`, instead of in a subdirectory of that

**Comment by dimbleby** [comment]
Ah, it's the trailing slash on `git = "https://github.com/instabase/opencensus-python/"` - that causes poetry's `Git.cls.get_name_from_source_url` to get the wrong answer, from which the rest follows.

Clearly either a more helpful error or just coping would be more helpful: but you can fix just by removing that trailing slash.

**Comment by LK** [comment]
Thanks for the quick debugging @dimbleby! That did seem to fix both this and #7326 as far as I can tell. Feel free to consolidate the two issues as you see fit since they seem to have the same root cause.

The trailing slash worked in our old pip-tools setup so was surprised to see it affect Poetry's behavior here when we migrated over — agreed that better handling of this case would be desirable (or at least preventing Poetry from bricking the venv when this happens).

**Comment by neersighted** [comment]
I don't think this is a duplicate -- I'm just going to consolidate into #7326 as this is a different root cause (so far there are 3: trailing slashes, mismatched names, and invalid Git repositories) for `.venv/src`-related issues.
